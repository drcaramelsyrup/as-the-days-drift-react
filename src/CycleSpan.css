.cycle {
	transition: all 0.3s ease-in-out;
}

.cycle[isInteractive="1"] {
	cursor: pointer;
	border-bottom: 1px rgba(211, 211, 211, 0.8) dotted;
	border-left: 1px solid rgba(25, 25, 122, 0.3);
	border-right: 1px solid rgba(25, 25, 122, 0.3);
}

.cycle[isInteractive="1"]:hover {
	color: #355c7e;
	border-left: 1px solid rgba(25, 25, 122, 0.0);
	border-right: 1px solid rgba(25, 25, 122, 0.0);
	border-bottom: 1px rgba(211, 211, 211, 0.0) dotted;
	font-size: 100.25%;
}

