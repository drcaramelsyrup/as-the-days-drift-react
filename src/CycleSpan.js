import React from 'react';
import './CycleSpan.css';

const CycleSpan = (props) => {
	const callback = props.callback != null ? props.callback : null;
	return isValidCycle(props.cycle) 
		&& (<span className='cycle' 
			onClick={ callback } 
			isInteractive={ isInteractive(props.cycle) }>{
				getCycleText(props.cycle)
			}</span>);
}

const isValidCycle = (cycle) => {
	return cycle != null 
		&& cycle.data != null 
		&& cycle.data.length > 0;
}

const isInteractive = (cycle) => {
	return isValidCycle(cycle) && cycle.data.length > 1
		? 1
		: 0;
}

const isValidCycleIdx = (cycleIdx) => {
	return cycleIdx != null && cycleIdx >= 0;
}

const getCycleText = (cycle) => {
	const cycleIdx = isValidCycleIdx(cycle.cycle_idx)
		? cycle.cycle_idx
		: 0;
	return cycle.data[cycleIdx].text;
}

export default CycleSpan;