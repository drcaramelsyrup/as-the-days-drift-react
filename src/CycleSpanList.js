import React from 'react';
import CycleSpan from './CycleSpan';
import { matchesQuery } from './InventoryUtils';
import { makeCycle, makeCycleData, updatedInventoryForCycle } from './CycleUtils';
import './CycleSpanList.css';

const CycleSpanList = (props) => {
	return <div className='CycleSpanList'>{
		props.data != null && getCycleSpans(
			props.data, 
			props.inventory,
			props.callback)	
	}</div>;
}

const getCycleSpans = (data, inventory = {}, callback = null) => {

  // TODO: bad practice to use idx in a component array
  const cycleArray = preprocessData(data, inventory);
  return cycleArray.reduce((acc, cycle, idx) => {
	return acc.concat([ <CycleSpan 
		key={ 'cyclespan' + idx } 
		cycle={ cycle }
		callback={ cycleSpanUpdateFunction(cycle, inventory, callback) }
	/> ]);
  }, []);

}

const resolveConditionals = (rawData, conditionals, inventory) => {

	if (conditionals == null)
		return rawData;

	const ids = conditionals.reduce((acc, conditional) => {
		if (matchesQuery(conditional.conditions, inventory))
			acc.relevant = acc.relevant.concat([ conditional.id ]);
		else
			acc.irrelevant = acc.irrelevant.concat([ conditional.id ]);
		return acc;
	}, { relevant: [], irrelevant: [] });

	const resolveTextWithConditional = (inText, conditionalId, conditionalText) => {
		if (conditionalId == null)
			return inText;
		if (!isValidString(conditionalText))
			return inText.replace(conditionalId, '');
		return inText.replace(conditionalId, conditionalText);
	}

	const resolvedData = Object.assign({}, rawData);

	const irrelevantIds = ids.irrelevant;
	resolvedData.text = irrelevantIds.reduce((text, id) => {
		return resolveTextWithConditional(text, id, '');
	}, resolvedData.text);

	const relevantIds = ids.relevant;
	resolvedData.text = relevantIds.reduce((text, id) => {
		// TODO: map this to optimize instead of running O(n) find by predicate?
		const conditionalIdx = conditionals.findIndex((cond) => { return cond.id === id; });
		const conditionalText = conditionalIdx >= 0
			? conditionals[conditionalIdx].text
			: '';
		return resolveTextWithConditional(text, id, conditionalText);
	}, resolvedData.text);

	return resolvedData;

}

const preprocessData = (rawData, inventory) => {
	
	const jsonData = resolveConditionals(rawData, rawData.conditionals, inventory);
	if (!isValidString(jsonData.text))
		return [];

	// Filter out cycles that do not appear and sort by order of appearance.
	const cycleIds = jsonData.cycles != null
		? Object.keys(jsonData.cycles)
			.filter(id => jsonData.text.indexOf(id) >= 0)
			.sort((a, b) => {
				return jsonData.text.indexOf(a) > jsonData.text.indexOf(b);
			})
		: [];

	// Only pure text, no cycles.
	if (cycleIds.length <= 0)
		return [ makeCycle(null, null, makeCycleData(jsonData.text)) ];

	const isValidIndex = (idx) => {
		return idx >= 0 && idx < cycleIds.length;
	}

	return cycleIds.reduce((acc, id, idx) => {

		// Start at the beginning or at the last cycle id position.
		const start = isValidIndex(idx - 1) 
			? jsonData.text.indexOf(cycleIds[idx - 1]) + cycleIds[idx - 1].length
			: 0;
		const end = isValidIndex(idx + 1)
			? jsonData.text.indexOf(cycleIds[idx + 1])
			: jsonData.text.length;

		const position = jsonData.text.indexOf(id);

		// Print before and after if it's the first, else print the after.
		const isFirstId = !isValidIndex(idx - 1);
		const before = isFirstId
			? jsonData.text.substr(start, position - start)
			: '';

		const after = jsonData.text.substr(
			position + id.length, end - position - id.length);

		const toConcat = [];

		if (before.length > 0)
			toConcat.push(makeCycle(null, null, makeCycleData(before)));

		const cycleData = jsonData.cycles[id];
		toConcat.push(makeCycle(id, getCycleIdx(id, cycleData, inventory), ...cycleData));

		if (after.length > 0)
			toConcat.push(makeCycle(null, null, makeCycleData(after)));

		return acc.concat(toConcat);

	}, []);
}

const getCycleIdx = (cycleId, cycleData, inventory) => {
	const desiredIdx = inventory.cycles != null ? inventory.cycles[cycleId] : 0;

	// Make sure that we have a valid cycle index.
	return getValidCycleIdx(cycleId, desiredIdx, cycleData, inventory);
}

const nextValidCycleIdx = (cycleId, currentIdx, cycleData, inventory) => {
	return getValidCycleIdx(cycleId, currentIdx + 1, cycleData, inventory);
}

const getValidCycleIdx = (cycleId, desiredIdx, cycleData, inventory) => {
	// Starting from the desired idx...
	if (desiredIdx == null)
		desiredIdx = 0;
	for (let i = desiredIdx; i < cycleData.length; ++i) {
		const datum = cycleData[i];
		if (matchesQuery(datum.conditions, inventory))
			return i;
	}
	for (let j = 0; j < desiredIdx; ++j) {
		const datum = cycleData[j];
		if (matchesQuery(datum.conditions, inventory))
			return j;
	}
	return -1;
}

// Returns false if string is empty, null, or undefined.
const isValidString = (str) => {
	return str && str.length > 0;
}

const cycleSpanUpdateFunction = (cycle, inventory, callback) => {
	return () => {
		const nextCycleIdx = nextValidCycleIdx(
			cycle.cycle_id, cycle.cycle_idx, cycle.data, inventory);
		if (nextCycleIdx === cycle.cycle_idx)
			return;

		callback(updatedInventoryForCycle(
			cycle.cycle_id, nextCycleIdx, cycle.data, inventory));
	}
}

export default CycleSpanList;
export { cycleSpanUpdateFunction };
