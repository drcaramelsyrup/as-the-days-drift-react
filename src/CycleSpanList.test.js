import React from 'react';
import ReactDOM from 'react-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer from 'react-test-renderer';
import CycleSpanList from './CycleSpanList';
import CycleSpan from './CycleSpan';
import data from './test_cyclespanlist.json';

describe('CycleSpanList - Smoke', () => {

	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<CycleSpanList />, div);
		ReactDOM.unmountComponentAtNode(div);
	});

});

describe('CycleSpanList - Render', () => {

	let renderer;

	beforeEach(() => {
		renderer = new ShallowRenderer();
	});

	afterEach(() => {
		renderer.unmount();
	});

	const makeExpectedCycleData = (actions, conditions, text) => {
		return {
			actions: actions,
			conditions: conditions,
			text: text
		};
	}

	const makeExpectedCycleSpan = (cycleId, cycleIdx, keyNum, ...data) => {
		return <CycleSpan key={ 'cyclespan' + keyNum } cycle={
			{ 
				cycle_id: cycleId, 
				cycle_idx: cycleIdx,
				data: [...data]
			}
		} />
	}

	it('renders as a div container', () => {
		renderer.render(<CycleSpanList />);
		const output = renderer.getRenderOutput();
		expect(output.type).toBe('div');
	});

	it('renders with empty text', () => {
		const dummyPassageData = { text: '' };
		renderer.render(<CycleSpanList data={ dummyPassageData } />);
		const output = renderer.getRenderOutput();
		expect(output.props.children).toEqual([]);
	})

	it('renders with simple text', () => {
		let text = 'test text';
		const dummyPassageData = {
			text: text 
		};

		renderer.render(<CycleSpanList data={ dummyPassageData } />);
		const output = renderer.getRenderOutput();
		const spanIdx = 0;
		expect(output.props.children).toEqual(
			[ makeExpectedCycleSpan(null, null, spanIdx,
				makeExpectedCycleData(null, null, text)) ]);
	});

	it('renders a simple cycling link', () => {
		const textBeforeLink = 'test before ';
		const testLinkText = 'this is a cycling link';
		const textAfterLink = ' test after';
		const cycleId = '_cycle_test';

		const dummyPassageData = {
			cycles: {
				[cycleId]: [
					{
						actions: null,
						conditions: null,
						text: testLinkText
					}
				]
			},
			text: textBeforeLink + cycleId + textAfterLink
		};

		renderer.render(<CycleSpanList data={ dummyPassageData } />);
		const output = renderer.getRenderOutput();

		let spanIdx = 0;
		const cycleIdx = 0;
		expect(output.props.children).toEqual([
			makeExpectedCycleSpan(null, null, spanIdx++,
				makeExpectedCycleData(null, null, textBeforeLink)),
			makeExpectedCycleSpan(cycleId, cycleIdx, spanIdx++,
				makeExpectedCycleData(null, null, testLinkText)),
			makeExpectedCycleSpan(null, null, spanIdx++,
				makeExpectedCycleData(null, null, textAfterLink))
		]);
	});

	it('renders a cycling link and can swap between cycles', () => {
		const firstLinkText = 'this is a cycling link';
		const secondLinkText = 'this is the second cycling link';
		const cycleId = '_cycle_test';
		const cycleIdx = 1;

		const dummyPassageData = {
			cycles: {
				[cycleId]: [{
					actions: null,
					conditions: null,
					text: firstLinkText
				}, {
					actions: null,
					conditions: null,
					text: secondLinkText
				}]
			},
			text: cycleId
		};

		renderer.render(<CycleSpanList data={ dummyPassageData } inventory={
			{
				cycles: {
					[cycleId]: cycleIdx
				}
			}
		} />);
		const output = renderer.getRenderOutput();

		const spanIdx = 0;
		expect(output.props.children).toEqual([
			makeExpectedCycleSpan(cycleId, cycleIdx, spanIdx, 
				makeExpectedCycleData(null, null, firstLinkText),
				makeExpectedCycleData(null, null, secondLinkText))
		]);
	});

	it('renders multiple cycling links in the same passage', () => {
		const firstCycleId = '_cycle_test_1';
		const firstText1 = 'test for cycle 1';
		const firstText2 = ' other test for cycle 1';

		const textBetween = ' ';

		const secondCycleId = '_cycle_test_2';
		const secondText1 = 'test for cycle2';

		const thirdCycleId = '_cycle_test_3';
		const thirdText1 = 'third';
		const thirdText2 = 'third cycle \n test ';

		const firstCycleIdx = 0;
		const secondCycleIdx = 0;
		const thirdCycleIdx = 1;

		const dummyPassageData = {
			cycles: {
				[firstCycleId]: [{
					actions: null,
					conditions: null,
					text: firstText1
				}, {
					actions: null,
					conditions: null,
					text: firstText2 
				}],
				// data positions should not affect ultimate placement
				[thirdCycleId]: [{
					actions: null,
					conditions: null,
					text: thirdText1
				}, {
					actions: null,
					conditions: null,
					text: thirdText2
				}],
				[secondCycleId]: [{
					actions: null,
					conditions: null,
					text: secondText1
				}]
			},
			text: firstCycleId + textBetween + secondCycleId + thirdCycleId
		};

		renderer.render(<CycleSpanList data={ dummyPassageData } inventory={
			{
				cycles: {
					[firstCycleId]: firstCycleIdx,
					[secondCycleId]: secondCycleIdx,
					[thirdCycleId]: thirdCycleIdx 
				}
			}
		} />);

		const output = renderer.getRenderOutput();
		let spanIdx = 0;
		const cycles = dummyPassageData.cycles;
		expect(output.props.children).toEqual([
			makeExpectedCycleSpan(firstCycleId, firstCycleIdx, spanIdx++, 
				makeExpectedCycleData(null, null, firstText1),
				makeExpectedCycleData(null, null, firstText2)),
			makeExpectedCycleSpan(null, null, spanIdx++,
				makeExpectedCycleData(null, null, textBetween)),
			makeExpectedCycleSpan(secondCycleId, secondCycleIdx, spanIdx++,
				makeExpectedCycleData(null, null, secondText1)),
			makeExpectedCycleSpan(thirdCycleId, thirdCycleIdx, spanIdx++,
				makeExpectedCycleData(null, null, thirdText1),
				makeExpectedCycleData(null, null, thirdText2))
		]);
	});

	describe('Conditions', () => {
		const cycleId = '_cycle_condition_test';
		const actions = {
			romantic: '1',
			complacent: '2'
		};
		const conditions = {
			neuroticism: {
				type: '<',
				val: '5'
			}
		};
		const conditionText = 'test condition';
		const testText = 'test render';
		const dummyPassageData = {
			pid: 0,
			cycles: {
				[cycleId]: [
					{
						actions: actions,
						conditions: conditions,
						text: conditionText
					},
					{
						actions: actions,
						conditions: null,
						text: testText
					}
				]
			},
			text: cycleId
		};

		const conditionalCycleIdx = 0;
		const nonconditionalCycleIdx = 1;
		const spanIdx = 0;

		const spanWithExpectedIdx = (cycleIdx) => {
			return makeExpectedCycleSpan(
				cycleId,
				cycleIdx,
				spanIdx,
				makeExpectedCycleData(
					actions, conditions, conditionText),
				makeExpectedCycleData(
					actions, null, testText));
		}

		it('does not render a span with conditions without an inventory', () => {
			renderer.render(<CycleSpanList data={ dummyPassageData } />);
			const output = renderer.getRenderOutput();
			expect(output.props.children).toEqual(
				[ spanWithExpectedIdx(nonconditionalCycleIdx) ]);
		});

		it('does not render a span with conditions when unsatisfied', () => {
			renderer.render(<CycleSpanList data={ dummyPassageData } inventory={
				{ neuroticism: 5 }
			} />);
			const outputNotSatisfied = renderer.getRenderOutput();
			expect(outputNotSatisfied.props.children).toEqual(
				[ spanWithExpectedIdx(nonconditionalCycleIdx) ]);
		});

		it('renders a span with conditions when satisfied', () => {
			renderer.render(<CycleSpanList data={ dummyPassageData } inventory={
				{ neuroticism: 3 }
			} />);
			const outputSatisfied = renderer.getRenderOutput();
			expect(outputSatisfied.props.children).toEqual(
				[ spanWithExpectedIdx(conditionalCycleIdx) ]);
		});
	});

	describe('Conditional Blocks', () => {
		const cycleId = '_cycle_condition_test';

		const firstConditionalId = '_cond_text_0';
		const secondConditionalId = '_cond_text_1';
		const firstConditionalText = '\nCome on, Sam\'s already inside. ';
		const secondConditionalText = 'He continued to fidget.';

		const dummyPassageData = {
			pid: 0,
			conditionals: [
		        {
		            conditions: {
		                partner: {
		                    type: 'is',
		                    val: 'sam'
		                },
		                var_nervous_habit: {
		                	type: 'equals',
		                	val: 'fidgeting'
		                }
		            },
		            id: '_cond_text_0',
		            text: firstConditionalText
		        },
		        {
		            conditions: {
		                romantic: {
		                    type: '<=',
		                    val: '2'
		                },
		                complacent: {
		                	type: '>=',
		                	val: '8'
		                }
		            },
		            id: '_cond_text_1',
		            text: secondConditionalText
		        }
		    ],
			cycles: {
				[cycleId]: [
					{
						actions: null,
						conditions: null,
						text: 'this text will never render'
					}
				]
			},
			text: firstConditionalId + secondConditionalId
		};

		it('does not render a conditional text block when unsatisfied', () => {
			const inventory = {	romantic: 5 };
			renderer.render(<CycleSpanList data={ dummyPassageData } inventory={ inventory } />);
			const output = renderer.getRenderOutput();
			expect(output.props.children).toEqual([]);
		});

		it('renders a conditional text block when satisfied', () => {
			const inventory = { 
				romantic: 6, complacent: 9, 
				partner: 'sam', var_nervous_habit: 'fidgeting' 
			};
			renderer.render(<CycleSpanList data={ dummyPassageData } inventory={ inventory } />);
			const output = renderer.getRenderOutput();
			const spanIdx = 0;
			expect(output.props.children).toEqual([
				makeExpectedCycleSpan(null, null, spanIdx,
					makeExpectedCycleData(null, null, firstConditionalText))
			])
		});

		it('renders a cycle span inside a conditional text block');
	});


	it('renders with a complete data passage', () => {
		const tree = TestRenderer.create(<CycleSpanList data={ data } />).toJSON();
		expect(tree).toMatchSnapshot();
	});

});

