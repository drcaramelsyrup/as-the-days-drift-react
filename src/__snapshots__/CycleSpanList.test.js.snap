// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CycleSpanList - Snapshot renders with a complete data passage 1`] = `
<div
  className="CycleSpanList"
>
  <span
    className="cycle"
    isInteractive={0}
    onClick={[Function]}
  >
    He was having trouble in 
  </span>
  <span
    className="cycle"
    isInteractive={1}
    onClick={[Function]}
  >
    precalculus, which was a problem since he wanted to be a civil engineer. You told him at this rate he'd only be building the sort of bridges that'd be in the front page of the papers, for all the wrong reasons
  </span>
  <span
    className="cycle"
    isInteractive={0}
    onClick={[Function]}
  >
    .
    
    You remember you bought him an organizer as a sort of light intervention. 
  </span>
  <span
    className="cycle"
    isInteractive={1}
    onClick={[Function]}
  >
    He'd still forget his homework at your house, a collection of unintelligible scribbles, and then forget to pick it up the next day
  </span>
  <span
    className="cycle"
    isInteractive={0}
    onClick={[Function]}
  >
    .
    
    Sorry, he'd said, $var_nervous_habit.
  </span>
</div>
`;
